src/main/scala/zio/metrics/MetricClient.scala

```diff
@@ -0,0 +1,179 @@
+ */
+// TODO: Rename that to MetricProxy
+// Change this to a Streaming backend builder
+// Also introduce the opposite like Batch backend builder
+// xx.build will produce a Layer building Unit
+// create zio.metrics.backends / NR / StatsD / Prometheus / ZIO
+//
+trait MetricClient {
+
+  /**
+   * Get the most recent snapshot. Tis method is typically used by backend
+   * implementations that require the current metric state on demand - such as
+   * Prometheus.
+   */
+  def snapshot(implicit trace: Trace): UIO[Set[MetricPair.Untyped]]
+
+  /**
+   * Register a new listener that can consume metrics. The most common use case
+   * is to push these metrics to a backend in the backend specific format.
+   */
+  def registerListener(listener: MetricListener[_])(implicit trace: Trace): UIO[Unit]
+
+  /**
+   * Deregister a metric listener.
+   */
+  // TODO: Rename to unregister
+  def deregisterListener(listener: MetricListener[_])(implicit trace: Trace): UIO[Unit]
+}
+
+object MetricClient {
+
+  def run = ZIO.serviceWithZIO[ZIOMetricClient](_.run)
+
+  def registerListener(l: MetricListener[_]) = ZIO.serviceWithZIO[MetricClient](_.registerListener(l))
+
+  def registerNewRelicListener() = for {
+    listener  <- ZIO.service[NewRelicListener]
+    publisher <- ZIO.service[NewRelicPublisher]
+    _         <- publisher.run
+    _         <- registerListener(listener)
+  } yield ()
+
+  final case class Settings(
+    pollingInterval: Duration)
+
+  object Settings {
+
+    val default = EnvVar
+      .duration("ZMX_METRIC_CLIENT_POLLING_INTERVAL", "MetricClient#Settings")
+      .getWithDefault(10.seconds)
+      .map(Settings(_))
+
+    val live = ZLayer.fromZIO(default)
+  }
+
+  class ZIOMetricClient private[MetricClient] (
+    settings: MetricClient.Settings,
+    listeners: Ref[Chunk[MetricListener[_]]],
+    latestSnapshot: Ref[Set[MetricPair.Untyped]])
+      extends MetricClient {
+
+    def deregisterListener(l: MetricListener[_])(implicit trace: Trace): UIO[Unit] =
+      listeners.update(cur => cur.filterNot(_.equals(l)))
+
+    def registerListener(l: MetricListener[_])(implicit trace: Trace): UIO[Unit] =
+      listeners.update(cur => cur :+ l)
+
+    def snapshot(implicit trace: Trace): UIO[Set[MetricPair.Untyped]] =
+      latestSnapshot.get
+
+    private def update(implicit trace: Trace): UIO[Unit] = for {
+      next       <- retrieveNext
+      registered <- listeners.get
+      _          <- ZIO.foreachPar(registered)(l => l.update(next))
+    } yield ()
+
+    private def retrieveNext(
+      implicit
+      trace: Trace,
+    ): UIO[Set[MetricEvent]] = for {
+      ts  <- ZIO.clockWith(_.instant)
+      res <- latestSnapshot.modify { old =>
+               // first we get the state for all the counters that we had captured in the last run
+               val oldMap = stateMap(old)
+               // then we get the snapshot from the underlying metricRegistry
+               val next   = metricRegistry.snapshot()
```

Comment:
Out of curiosity, why do you periodically call the snapshot method as opposed to registering your own listener in the `metricRegistry` and sending the individual events to your queue?

Reply:
Hi there, we are moving to a completely pull  based API and will get rid of the metric listeners. The thought behind that is that even for the "push" based clients like StatsD or New Relic we have agents that sit between the monitored application and the backend. These will accumulate and report the metrics at certain intervals. 

With getting rid of the MetricListener we are streamlining the API.